### PR TITLE
Handling mixed precision for dreambooth flux lora training

### PR DESCRIPTION
**Pull request by icsl-Jeon**
# What does this PR do?
Hello 😄 Thank you for the awesome example!
Here, I want to make a PR that helped me train dreambooth LoRA successfully.

- Avoid `dtype` change for `transfermer` after `log_validation` (especially for `fp16`). For mixed training, the original code [upcast fp16 to fp32](https://github.com/huggingface/diffusers/blob/33fafe3d143ca8380a9e405e7acfa69091d863fb/examples/dreambooth/train_dreambooth_lora_flux.py#L1277) for mixed precision training.  However, after switching `pipeline` dtype in `log_validation`, `transformer` dtype returns to `fp16`, which can lead to [fp16 unscaling error.](https://github.com/huggingface/diffusers/issues/6442) Actually, I had this problem when I use `fp16`  option. (For some reason, T5 yielded nan output in bf16, that is why I came to use fp16) https://github.com/huggingface/diffusers/issues/9476
- Use the same preset `dtype` for text encoders for validation. For some reason, the two text encoders were in fp32 in the `pipeline`.  
- Add `unwarp` to access the `config` field of `transformer`. https://github.com/huggingface/diffusers/issues/9161#issuecomment-2297897162

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [x] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

[linoytsaban](https://github.com/linoytsaban) @sayakpaul

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: examples/dreambooth/train_dreambooth_lora_flux.py [hunk 1]
```diff
@@ -177,7 +177,7 @@ def log_validation(
         f"Running validation... \n Generating {args.num_validation_images} images with prompt:"
         f" {args.validation_prompt}."
     )
-    pipeline = pipeline.to(accelerator.device, dtype=torch_dtype)
+    pipeline = pipeline.to(accelerator.device)
```
Comment (sayakpaul) [inline examples/dreambooth/train_dreambooth_lora_flux.py:180]: Why are we doing it? 

We should keep the pipeline model-level components (such as text encoders, VAE, etc.) to a reduced precision no?

**Review by sayakpaul — COMMENTED** [review]
Thank you! Just a single comment. 

@linoytsaban could you also give this a look?

**Comment by sayakpaul** [comment]
> Avoid dtype change for transfermer after log_validation (especially for fp16). For mixed training, the original code [upcast fp16 to fp32](https://github.com/huggingface/diffusers/blob/33fafe3d143ca8380a9e405e7acfa69091d863fb/examples/dreambooth/train_dreambooth_lora_flux.py#L1277) for mixed precision training. However, after switching pipeline dtype in log_validation, transformer dtype returns to fp16, which can lead to https://github.com/huggingface/diffusers/issues/6442 Actually, I had this problem when I use fp16 option. (For some reason, T5 yielded nan output in bf16, that is why I came to use fp16)

We could avoid it by running inference in autocast, no? Here's an example:
https://github.com/huggingface/diffusers/blob/61d37640ade33f1e2d330a51466b94dfe4f155f6/examples/dreambooth/train_dreambooth_lora_sdxl.py#L215

File: examples/dreambooth/train_dreambooth_lora_flux.py (hunk 1 above)
Comment (icsl-Jeon) [reply examples/dreambooth/train_dreambooth_lora_flux.py:180]: text encoders, vae are already in reduced precision :) 
As I described in the PR description, this will change dtype of `transformers` 
For mixed precision training, transformer was upcast into fp32 if fp16 training.
But this changes back to fp16, which leads to fp16 unscale error in clip gradient.

Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_flux.py:180]: Would something like this work? 
https://github.com/huggingface/diffusers/pull/9549#issuecomment-2388677554

Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_flux.py:180]: Ah okay. Can you provide an example command for us to verify this? Maybe @linoytsaban could give it a try?

Comment (sayakpaul) [reply examples/dreambooth/train_dreambooth_lora_flux.py:180]: @icsl-Jeon a friendly reminder :)

Comment (icsl-Jeon) [reply examples/dreambooth/train_dreambooth_lora_flux.py:180]: This could be reproduced with any launch commands in the [README](https://github.com/huggingface/diffusers/blob/main/examples/dreambooth/README_flux.md).
`accelerate launch ... --mixed_precision="fp16" ..`

I checked the lora precision with 
```
            for name, param in transformer.named_parameters():
                if 'lora' in name:
                    print(f"Layer: {name}, dtype: {param.dtype}, requires_grad: {param.requires_grad}")
```
Hope this help you reproduce!

**Review by sayakpaul — APPROVED** [review]
LGTM actually! 

@linoytsaban could you also review?

**Comment by icsl-Jeon** [comment]
@linoytsaban thank you in advance

**Review by linoytsaban — APPROVED** [review]
Thanks @icsl-Jeon, LGTM!

**Comment by icsl-Jeon** [comment]
Is there any action to be done for merge ?

**Comment by sayakpaul** [comment]
Will merge this once the CI run is complete. Thanks a ton!

**Comment by sayakpaul** [comment]
Thanks for your contributions!
